Switch to using ParseField to parse query names

javanna opened the pull request:
If we have a deprecated query name, at the moment we don't have a way to log any deprecation warning. With this change we use ParseField, so that we can for instance deprecate the "in" variant of the terms query etc. and output proper deprecation warnings.

javanna commented:
This does not even compile at the moment, it is just a prototype to share with @colings86 @nik9000 and @cbuescher can have a look before I go ahead with the change. I moved only the newer registration method to ParseField to minimize the work at this first stage.

javanna commented on core/src/test/java/org/elasticsearch/search/SearchModuleTests.java:
```diff
@@ -82,6 +83,98 @@ public void testRegisterQueryParserDuplicate() {
 
+    public void testRegisteredQueries() {
+        //TODO rework test
+        /*SearchModule module = new SearchModule(Settings.EMPTY, new NamedWriteableRegistry());
```
I will fix this test once we settle on how things should look, but the idea is to have here a test that verifies which queries get registered, and duplicate all the query names. I think this is the only safe way to make sure that nothing gets lost.

colings86 commented on core/src/main/java/org/elasticsearch/indices/query/IndicesQueriesRegistry.java:
```diff
@@ -19,24 +19,47 @@
+        assert match : "registered parse field did not match the query name it was registered for";
+        QueryParser<?> queryParser = queryParsers.get(name);
+        if (queryParser == null) {
+            throw new ParsingException(xContentLocation, "No parser registered for [" + name + "]");
```
Should this also be an assert since it should not be possible to register a parseField without a queryParser?

javanna replied:
yea probably

colings86 commented on core/src/main/java/org/elasticsearch/search/SearchModule.java:
```diff
@@ -392,15 +396,22 @@ public IndicesQueriesRegistry buildQueryParserRegistry() {
         for (QueryRegistration<?> query : queries) {
             QueryParser<? extends QueryBuilder<?>> parser = query.parser;
-            for (String name: query.names) {
+            for (String name: query.parseField.getAllNamesIncludedDeprecated()) {
```
Maybe it's too much for this PR but would it be better for IndicesQueryRegistry to have a `registerQuery(ParserField, QueryParser)` method which then makes sure it builds the parseField and queryParsers maps properly internally?

javanna replied:
maybe, I would also like to simplify things, cause at the moment we internally need QueryRegistration as we delay the actual registration to when buildQueryParserRegistry gets called. I may try this once I migrate the other register method. Or give up if it gets too big as part of this PR.

nik9000 replied:
I dropped QueryRegistration in https://github.com/elastic/elasticsearch/pull/17458.

colings86 replied:
Sounds good. I was also wondering if we need the QueryRegistration object since I think we should be able to directly register the query with the named writeable registry and the parser registry in the register method

colings86 replied:
@nik9000 awesome :smile:

colings86 commented:
@javanna I left a couple of comments but I like it

cbuescher commented on core/src/main/java/org/elasticsearch/search/SearchModule.java:
```diff
@@ -392,15 +396,22 @@ public IndicesQueriesRegistry buildQueryParserRegistry() {
+                }
+                oldValue = parseFieldsMap.putIfAbsent(name, query.parseField);
+                if (oldValue != null) {
+                    throw new IllegalArgumentException("Query parser [" + oldValue + "] already registered for name [" + name +
```
nit: Query parse field?

cbuescher commented on core/src/main/java/org/elasticsearch/indices/query/IndicesQueriesRegistry.java:
```diff
@@ -19,24 +19,47 @@
+    private ParseFieldMatcher parseFieldMatcher;
 
-    public IndicesQueriesRegistry(Settings settings, Map<String, QueryParser<?>> queryParsers) {
+    public IndicesQueriesRegistry(Settings settings, Map<String, QueryParser<?>> queryParsers, Map<String, ParseField> parseFields) {
```
It would be nice if we could somehow get the parseField from each parser, but I guess the problem is that QueryParser needs to be a functional interface? Maybe this can be added later.

javanna replied:
yes that's the reason. Parsers will be gone soon apart from the functional interface.

nik9000 replied:
It doesn't _need to_ be a functional interface. It makes it easier to work with, I think.

nik9000 replied:
`Map<String, Tuple<ParseField, QueryParser<?>>>`?

javanna replied:
I am all for adding this new parameter, that is what I've been doing. I don't love tuples, I think they would complicate things even more here.

javanna replied:
Sorry I understood only later what Nik meant, I am changing to Tuple instead of two maps. thanks for pointing that out. I thought we were talking about changing the fromXContent method of the functional interface to something that would use Tuple... but using Tuple internally is fine and actually saves me from some synchronization issues that need to be handled is we have two maps.

cbuescher commented:
I also think this looks good, two minor comments, the second is also just idea maybe for future improvements, don't know if that would work currently.

javanna commented on core/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:
```diff
@@ -63,6 +63,10 @@
     /** The default name for the match query */
     public static final String NAME = "match";
 
+    //TODO all these alternate names are now deprecated but the MatchQueryBuilder code needs to be updated.
```
this is for #17508 which I think should go in first. otherwise we deprecate before having an alternative.

javanna commented on core/src/main/java/org/elasticsearch/index/query/ScriptQueryParser.java:
```diff
@@ -65,7 +61,8 @@ public ScriptQueryBuilder fromXContent(QueryParseContext parseContext) throws IO
                     script = Script.parse(parser, parseContext.parseFieldMatcher());
-                } else if (parseContext.parseFieldMatcher().match(currentFieldName, PARAMS_FIELD)) { // TODO remove in 3.0 (here to support old script APIs)
+                } else if (parseContext.parseFieldMatcher().match(currentFieldName, PARAMS_FIELD)) {
+                    // TODO remove in 3.0 (here to support old script APIs)
```
this TODO is not good :( but it was there already. We should probably remove this stuff in another PR.

javanna commented:
I updated the PR and removed the WIP label. I addressed all comments, don't get fooled by the size, it's the same small change all over the place, plus some line length problems addressed while I was at it. Most of the change is in SearchModule, where I made some of the changes that Nik made as part of #17458. We also have a test for registered queries in SearchModuleTests, let me know what you think!

nik9000 commented on docs/reference/query-dsl/mlt-query.asciidoc:
```diff
@@ -6,7 +6,7 @@ set of documents. In order to do so, MLT selects a set of representative terms
 and returns the results. The user controls the input documents, how the terms
 should be selected and how the query is formed. `more_like_this` can be
-shortened to `mlt`.
+shortened to `mlt` (deprecated, use `more_like_this` rather than `mlt`).
```
Can this have a version attached?

javanna replied:
yea I'll have to figure out how to do that, I forgot

nik9000 commented:
I'm fine with the code. We'll need to add stuff to the breaking changes docs I think. As well as put the versions when things were deprecated in the docs.

cbuescher commented:
I took a look, this looks good to me.

javanna commented:
> We'll need to add stuff to the breaking changes docs I think.

but nothing is breaking here :) I am not sure whether we have a deprecation list somewhere, I will ask around.

nik9000 commented:
The breaking changes list is the deprecated list, iirc.

colings86 commented on core/src/main/java/org/elasticsearch/search/SearchModule.java:
```diff
@@ -305,20 +312,39 @@ public void registerFunctionScoreParser(ScoreFunctionParser<? extends ScoreFunct
+    }
+
+    private <QB extends QueryBuilder<QB>> void innerRegisterQueryParser(QueryParser<QB> parser, ParseField queryName) {
+        Tuple<ParseField, QueryParser<?>> parseFieldQueryParserTuple = new Tuple<>(queryName, parser);
```
To me, this logic should really be in `IndicesQueriesRegistry` so we construct the registry with just the `Settings` object and then call a `registerQuery(ParseField, QueryParser<?>)` method which unpacks the `ParseField` and adds it to the registry map. That was the registry is dealing with how to internally structure the data and the internals can be easily changed later without affecting outside code.

colings86 replied:
To me, this logic should really be in `IndicesQueriesRegistry` so we construct the registry with just the `Settings` object and then call a `registerQuery(ParseField, QueryParser<?>)` method which unpacks the `ParseField` and adds it to the registry map. That was the registry is dealing with how to internally structure the data and the internals can be easily changed later without affecting outside code.

colings86 replied:
To me, this logic should really be in `IndicesQueriesRegistry` so we construct the registry with just the `Settings` object and then call a `registerQuery(ParseField, QueryParser<?>)` method which unpacks the `ParseField` and adds it to the registry map. That was the registry is dealing with how to internally structure the data and the internals can be easily changed later without affecting outside code.

javanna replied:
I get it I get it :) I think it is easy to move this method to the registry at this point, since we have a single step only

javanna replied:
I changed my mind, I think the register methods should be only here in SearchModule. The nice thing of IndicesQueriesRegistry is that it's immutable, and classes that don't depend on it do so only because they need to read from it, retrieve registered parsers. I wouldn't want to expose a register method there and make the class mutable. You see what I mean?

colings86 replied:
Hmmm, I do see what you mean. Personally I would still prefer the register method in the registry but I think this is a personal preference thing rather than a substantial concern so I'm happy to yield on it :smile:

colings86 commented:
@javanna I added one more comment but otherwise it LGTM

colings86 commented:
LGTM
